### PR TITLE
Revert an incorrect optimization in binary operator overload resolution

### PR DESCRIPTION
**Pull request by gafter**
from PR #9543; issue #13380 shows why the optimization is not correct.

@dotnet/roslyn-compiler Please review
/cc @VSadov @tmat @agocke @AlekseyTs

### CODE REVIEW

**Comment by gafter** [comment]
@VSadov @dotnet/roslyn-compiler Please review. This is not a large change.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs [hunk 1]
```diff
@@ -843,21 +811,17 @@ private static LiftingResult UserDefinedBinaryOperatorCanBeLifted(TypeSymbol lef
...
                 {
-                    if (i == j)
-                    {
-                        continue;
-                    }
-                    if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
+                    if (candidates[j].Kind != OperatorAnalysisResultKind.Applicable)
                     {
                         continue;
                     }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs]: Can the same changes be made to `UnaryOperatorOverloadResolution`?

**Comment by VSadov** [comment]
LGTM

We should think about how to get back the benefits of this optimization without breaking. This code causes a noticeable amount of allocations while the vast majority of cases are trivial and could be handled without involving nullable enums ets...

File: src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs]: Actually, I'm not sure this change is correct. It is conceivable that betterness is not transitive, and we have a cycle such that M1 is better than M2, M2 is better than M3, and M3 is better than M1. In such a case this change is not correct (no candidate is better than all others). Similarly, we could have M1 better than M2, M2 better than M3, but not M1 better than M3. In that case depending on the order that the methods appear in the array, we could get different answers (either ambiguous or M1 is best). So I'm going to revert the change on line 824.

However, the change to the loop indices appears correct and can probably be made elsewhere if applicable.

---
In reply to: [78035627](https://github.com/dotnet/roslyn/pull/13648#discussion_r78035627) [](ancestors = 78035627)

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs]: On the other hand, if BetterOperator is a transitive property (which would need to be proven) then this optimization is correct, and in fact the algorithm can probably be simplified further to the linear-time algorithm hinted at in the comment above.

---
In reply to: [78047238](https://github.com/dotnet/roslyn/pull/13648#discussion_r78047238) [](ancestors = 78047238,78035627)

**Comment by cston** [comment]
LGTM, assuming we have the same optimizations in `Unary/BinaryOperatorOverloadResolution` loops.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs]: So I'm either going to prove that it is transitive (and then add comments to that effect) or revert line 824.

---
In reply to: [78049650](https://github.com/dotnet/roslyn/pull/13648#discussion_r78049650) [](ancestors = 78049650,78047238,78035627)

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs]: I believe it is not transitive, so I'm reverting 824.

---
In reply to: [78060790](https://github.com/dotnet/roslyn/pull/13648#discussion_r78060790) [](ancestors = 78060790,78049650,78047238,78035627)

**Comment by gafter** [comment]
I have filed #13683 to track getting this optimization restored.

---
In reply to: [245664465](https://github.com/dotnet/roslyn/pull/13648#issuecomment-245664465) [](ancestors = 245664465)

**Comment by gafter** [comment]
A forthcoming iteration will add that.

---
In reply to: [245672859](https://github.com/dotnet/roslyn/pull/13648#issuecomment-245672859) [](ancestors = 245672859)

**Comment by gafter** [comment]
I added an iteration to (1) remove a possibly incorrect "optimization, and (2) port a small loop optimization from the Binary case to the Unary case (it halves the cost of selecting the best candidate).

**Comment by gafter** [comment]
I'll hold this PR open until tomorrow in case you have any further comments on the latest iteration.

**Pull request merged**